src/app/copy/block-copy/block-copy.component.html

```diff
@@ -0,0 +1,33 @@
+<div>
+  <label [attr.for]="'pf-copy-token-button-' + hash">{{label}}</label>
+  <div class="copy">
+    <div class="copy-preview" [ngClass]="{'pf-is-open': tokenPanelOpen}">
+
+      <button
+        [attr.aria-label]="expandBtnAriaLabel(label)"
+        [attr.aria-expanded]="tokenPanelOpen"
+        class="copy-preview-btn"
+        (click)="toggleTokenPanel()">
+        <i
+          class="fa copy-preview-icon"
+          [ngClass]="{'fa-angle-down': tokenPanelOpen, 'fa-angle-right': !tokenPanelOpen}"></i>
+      </button>
+
+      <div class="copy-preview-txt-cont">
+        <span
+          data-toggle="tooltip"
+          data-placement="top"
+          [tooltip]="label">{{token}}</span>
+      </div>
+
+      <button
+        [attr.id]="'pf-copy-token-button-' + hash"
```

Comment:
Is the hash being used to create a unique ID? If so, we should talk about standardizing how we do that for patternfly-ng components. We may want to allow the application to provide its own ID.

FYI, there is an issue to provide checkbox IDs for pfng-list. Need to make sure whatever we do here is consistent.

Reply:
@dongniwang or @jgiardino may know of a tie breaker. Definitely worth checking if there's a single solution for both this and pfng-list.

Reply:
Sure, so the flow would be; a user clicks on the text label for the block copy component, which causes the text to be copied to their clipboard.

In general, for form controls, it's expected that when a user clicks the label, it will result in the same behavior as if the user clicks the corresponding form control directly. The more common scenario is when a user clicks a label resulting in focus placed in a related input. With the related control being a button, though, the result is whatever happens when you click the button.

Would this be considered unexpected behavior? I'm starting to think it may be, but wanted another opinion. It's easy to make this work either way, just wondering what's best.

Noticed the focus leaving the button on keypress, I'll fix this. Good catch.

Reply:
My concern is more about how do we generate IDs for patternfly-ng overall. I'm thinking if the user provides an ID, we should use that as a prefix. Lodash has a uniueid method we could use by default if an ID isn't given.

Reply:
If we were to use a provided "id" as a prefix, then it becomes, well, not the id anymore. I'm wondering what value this would add from a users perspective. It shouldn't really matter since it's a generated id anyway, right? I could be missing something..

I'm starting to think uniqueId primed with "pfng-block-copy-button-" does the job well enough and is inline with the the work you did over in list recently - https://github.com/patternfly/patternfly-ng/pull/360/

One caveat - I appended an extra dash to the prefix in my latest changes because I think it looks nice with the counter being separated with a dash like the rest of the string, but I could be convinced otherwise :)

@dlabrecq let me know what you think, happy to provide a user defined id prefix, just really wanna understand what this might be used for before I put it in there. Thx!

Reply:
Being able to provide an ID allows users to write tests based on something that is known to them. For example, if someone wrote a Selimium test to activate the button, the test would need to rely on a consistent ID. A unique, generated ID isn't something a test can rely on because it is always different. Using a generated ID is fine by default, but we need to allow for that to be overridden. The convention used by the components should be consistent -- please follow #360.